encode_wav/src/lib.rs

```diff
@@ -168,25 +168,28 @@ impl<B: WavBitDepth + 'static> Encoder for WavEncoder<B> {
                     .write_to_disk(&write_block.block()[0][0..written_frames], &mut file)?;
             } else {
                 if self.num_channels == 2 {
-                    // Hint to compiler to optimize loop
-                    assert!(written_frames <= write_block.block()[0].len());
-                    assert!(written_frames <= write_block.block()[1].len());
-                    assert!(written_frames * 2 <= self.interleave_buf.len());
-
-                    for frame in 0..written_frames {
-                        self.interleave_buf[frame * 2] = write_block.block()[0][frame];
-                        self.interleave_buf[frame * 2 + 1] = write_block.block()[1][frame];
+                    // Provide efficient stereo interleaving.
+                    let ch1 = &write_block.block()[0][0..written_frames];
+                    let ch2 = &write_block.block()[1][0..written_frames];
+                    let interleave_buf_part = &mut self.interleave_buf[0..written_frames * 2];
+
+                    for (i, frame) in interleave_buf_part.chunks_exact_mut(2).enumerate() {
```

Comment:
Instead of enumerating and indexing `ch1` and `ch2` with `i` you could also zip them which may elide further bound checks, but perhaps the compiler already takes care of this

Reply:
I checked the assembly output in release mode, and surprisingly the indexing version is more optimized than the zip version. So I'll keep it as it is.

If you're curious, this is the assembly output of the loop with indexing (or at least the part of the loop that makes use of SIMD intrinsics which is where the bulk of the work is done):
```assembly
.LBB0_10:
        movsd   xmm0, qword ptr [rax + 4*r10]
        movsd   xmm1, qword ptr [rdx + 4*r10]
        unpcklps        xmm0, xmm1
        movups  xmmword ptr [r8 + 8*r10], xmm0
        add     r10, 2
        cmp     rsi, r10
        jne     .LBB0_10
```

And this is with using zip:
```assembly
.LBB1_7:
        movsd   xmm0, qword ptr [rdi + 4*rcx]
        movsd   xmm1, qword ptr [rdi + 4*rcx + 8]
        movsd   xmm2, qword ptr [rdx + 4*rcx]
        unpcklps        xmm0, xmm2
        movsd   xmm2, qword ptr [rdx + 4*rcx + 8]
        unpcklps        xmm1, xmm2
        movups  xmmword ptr [r8 + 8*rcx], xmm0
        movups  xmmword ptr [r8 + 8*rcx + 16], xmm1
        add     rcx, 4
        cmp     rax, rcx
        jne     .LBB1_7
```

Auto-vectorizing interleaving/deinterleaving is kind of hard for a compiler to do.

Reply:
Actually, that gives me the idea of creating a crate for efficient interleaving/deinterleaving. I'll think I'll do that.

Reply:
Interesting!